### PR TITLE
SocketsHttpHandler: Ensure that we don't initiate connection creation under the pool lock

### PR DESCRIPTION
**Pull request by geoffkizer**
This change moves the connection initiation outside of the pool lock.

Also, refactor slightly so that there is a single path for new connection creation, whether this happens because we are below the connection limit, or when the waiter completes with an available connection count but no actual connection.

@stephentoub

### CODE REVIEW

**Comment by geoffkizer** [comment]
@dotnet-bot test outerloop linux x64 debug build
@dotnet-bot test outerloop Windows x64 debug build
@dotnet-bot test outerloop OSX x64 debug build

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 1]
```diff
@@ -210,57 +210,52 @@ private static SslClientAuthenticationOptions ConstructSslOptions(HttpConnection
             return GetHttpConnectionAsync(request, cancellationToken);
         }
 
-        private ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)> 
-            GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        private ValueTask<HttpConnection> GetOrReserveHttp11ConnectionAsync(CancellationToken cancellationToken)
         {
             if (cancellationToken.IsCancellationRequested)
             {
```
[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Yes, on purpose. One thing I've noticed in using tracing recently is that it's too verbose about trivial things, e.g. objects getting created. This trace seemed rather trivial, so I removed it.

That said, looking at this particular instance, this is not really an example of unnecessary verbosity; I may have been too eager here.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 2]
```diff
@@ -289,18 +284,39 @@ private static SslClientAuthenticationOptions ConstructSslOptions(HttpConnection
...
+                if (connection == null)
+                {
+                    Debug.Assert(failureResponse != null);
+                    DecrementConnectionCount();
+                }
+                return (connection, true, failureResponse);
+            }
+            catch
+            {
+                DecrementConnectionCount();
```
[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: DecrementConnectionCount shouldn't throw. That said, I agree that it's better to restructure the code as you suggest. I'll revisit this.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test NETFX x86 Release Build
Build please
@dotnet-bot test Windows x64 Debug Build
Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Pull request merged**